### PR TITLE
Scan block transactions in parallel

### PR DESCRIPTION
**Pull request by romanz**

### CODE REVIEW

File: src/status.rs [hunk 1]
```diff
@@ -353,13 +367,27 @@ impl ScriptHashStatus {
...
             .collect();
         self.for_new_blocks(spending_blockhashes, daemon, |blockhash, block| {
-            let txids: Vec<Txid> = block.txdata.iter().map(|tx| tx.txid()).collect();
-            for (pos, (tx, txid)) in block.txdata.into_iter().zip(txids.iter()).enumerate() {
-                let spent_outpoints = filter_inputs(&tx, outpoints);
-                if spent_outpoints.is_empty() {
-                    continue;
-                }
-                cache.add_tx(*txid, move || tx);
+            let txids: Vec<Txid> = block.txdata.par_iter().map(|tx| tx.txid()).collect();
```
[the remark this answers is not in the document]
Comment (romanz) [reply src/status.rs:370]: Not sure I understand... do you mean that `txids` are computed twice?

File: src/status.rs [hunk 2]
```diff
@@ -331,13 +331,27 @@ impl ScriptHashStatus {
...
+            let items: Vec<(usize, (Transaction, &Txid))> = block
+                .txdata
+                .into_iter()
+                .zip(txids.iter())
+                .enumerate()
+                .collect();
+
+            let found: Vec<(usize, &Txid, Vec<TxOutput>)> = items
```
[the remark this answers is not in the document]
Comment (romanz) [reply src/status.rs:342]: Good point, thanks!

File: src/status.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Kixunil) [reply src/status.rs:366]: Yes

[the remark this answers is not in the document]
Comment (romanz) [reply src/status.rs:366]: I am not sure... `txids` [here](https://github.com/romanz/electrs/pull/597/files#diff-0bfaab7c77d86eb41c2fe32c5e5df509fa02ab04c97b0ce3216911afaae0d67cR366) are computed for each "spending" block, while the `txids` [above](https://github.com/romanz/electrs/pull/597/files#diff-0bfaab7c77d86eb41c2fe32c5e5df509fa02ab04c97b0ce3216911afaae0d67cR334) are computed for each "funding" block. I think that for most addresses, funding and spending transactions are usually in different blocks - so I am not sure that we would gain much by caching `txids` (IIUC the suggestion).
In addition, I am not sure that computing `txids` is the bottleneck here :(

[the remark this answers is not in the document]
Comment (romanz) [reply src/status.rs:366]: I can rename `txids` to `spending_txids` / `funding_txids` to make the code more readable... WDYT?

[the remark this answers is not in the document]
Comment (Kixunil) [reply src/status.rs:366]: Ah, I see. Yes, renaming is a good idea.

[the remark this answers is not in the document]
Comment (romanz) [reply src/status.rs]: Renamed in https://github.com/romanz/electrs/pull/597/commits/3ce23ae01e44ed6183fa95d892953e70a22c8c88.

**Comment by Kixunil** [comment]
One more improvement I can see:

0. write `fn for_new_blocks_par(...) -> impl ParallelIterator<Item=Result<Block, Error>>`
1. implement funding txes pass roughly as `for_new_blocks_par(...).flat_map(|block| block.transactions).filter_map(...).try_collect()?;` The else branch in filter adds the transaction into `new_transactions` which is some collection that's cheap to build in parallel (lock-free linked list?) and can be turned into `ParallelIterator`
2. Implement spending txes pass as iteration on top of `ParallelIterator` created from `new_transactions`

This would avoid fetching blocks twice, avoid the possible overhead of calling `par_iter()` multiple times and improve parallelization opportunities.

**Comment by Kixunil** [comment]
Actually adding probably has to be unconditional, will have to think about it but I have to go now.

**Comment by romanz** [comment]
I refactored the above code in https://github.com/romanz/electrs/commit/fefd9dee11d19b682eb0e9bb929c89f7a1b782e1 and https://github.com/romanz/electrs/commit/c8e41c4052859e20f97361b0f7d0f376ec761440, please take a look :)

**Pull request merged**